### DISCUSSION TITLE
Vector map load performance slower after MapLibre version 3.0.0-pre.5?

### DISCUSSION THREAD

**Discussion by Al-4SW** [maplibre/maplibre-gl-js#4674] [Q&A] [answered]
I've just update from using Maplibre v2.4.0 to v4.7.0 and noticed that vector map loading appears slower than in v2.4.0.

On v 2.4.0, whilst zooming in/out, tiles appeared very smoothly. After doing some testing on previous versions today, it seems after mapLibre v3.0.0-pre.5, map tiles take longer to load when zooming in/out. 

With 3.0.0-pre.6 and later, when zooming in I get a partially loaded tile with 'jagged' roads and not all layers. When the zoom event ends, I get the full tile, but the delay is quite noticeable vs older versions. Similar for zooming out - empty areas before tiles are loaded seem a lot more noticeable on newer versions.

 I've tried different browsers and map tile providers and it seems pretty consistent, at least for me anyway, that map tiles are loading in significantly quicker & smoother with release prior to v3.0.0-pre.5? I've tried setting the mapOptions cancelPendingTileRequestsWhileZooming to false, but it doesn't seem to make any difference?

has anyone else experienced this? Are there any settings I can change  to get closer to 2.4 performance?

**Comment by HarelM** [comment]
"Feels shower" isn't a great way to measure things...
I would advise to find a better way to measure degradation in performance. 
I would also suggest to go over the changelog and see which PRs might have caused this issue and see if reverting them helps. 
The main zoom related breaking change is https://github.com/maplibre/maplibre-gl-js/pull/2377, which later a flag was added to allow reverting this behavior, as you mentioned. 
I'm not saying there's no issue, but I believe you'll need to investigate what can solve this in latest version according to the changelog PRs.

**Reply by Al-4SW** [reply, marked as answer]
Hi @HarelM , thanks for pointing me in the right direction.  I have found the cause for difference in performance I was experiencing between versions.

V2.4.0 defaults to 6 workers, #2354 introduced default of 1 worker in V3.0.0-pre.6.

Setting maplibregl.setWorkerCount(6) in latest v4.7.0 , I now get similar tile load performance to v2.4.0.

**Reply by HarelM** [reply]
The worker change is only applied for iOS as far as I remember, can you attach a video showing the difference by any chance?
If this is a performance degradation we might consider reverting this as part of version 5 breaking changes.
CC: @zhangyiatmicrosoft

**Reply by Al-4SW** [reply]
MapLibre v2.4.0 all default values:
https://github.com/user-attachments/assets/15012b0c-f3e2-4d54-a30a-fe43bc6bc5fa

MapLibre v4.7.0 all default values:
https://github.com/user-attachments/assets/67127fbf-ae63-4247-81c6-5e61c80ed649

MapLibre v4.7.0 with 'maplibregl.setWorkerCount(6)' :
https://github.com/user-attachments/assets/198faffb-8d75-4a12-9ba2-aa04970b25dd

All examples above are on Firefox 130.0 64 bit Linux. I've found chrome has similar results. cancelPendingTileRequestsWhileZooming is left at default true for the 4.7.0 examples.

 I've also tried experimented with different workerCount values in 4.7.0. Even just increasing to 2 I see reasonable improvement in the 'progressive' tile load performance.

Somewhat off topic, I'm not sure if it's apparent in the video uploads, but the mouse scroll is not as smooth in 4.7.0 as the 2.4.0 example. I also experience similar stutter issues as others have noted in recent comments to #4362.

**Reply by HarelM** [reply]
It should only be 1 for safari, see here: 
https://github.com/maplibre/maplibre-gl-js/blob/41e5b32f5bd6264cbc4a8b38210ec6a410152259/src/util/worker_pool.ts#L58

What does this line of code returns in your case?

**Reply by Al-4SW** [reply]
It's the other way around; it's setting min 1 max 3 for Safari (depending on available cores reported), 1 for all other browsers.

 I did run it anyway and logged the value, I got 1.

**Reply by HarelM** [reply]
Oh, right, My mistake...
This is the relevant PR:
https://github.com/maplibre/maplibre-gl-js/pull/2354
